source/MaterialXContrib/MaterialXNode/MaterialXData.cpp

```diff
@@ -1,49 +1,93 @@
-#include <MaterialXData.h>
-#include <Util.h>
+#include "MaterialXData.h"
+#include "MaterialXUtil.h"
 
 #include <MaterialXFormat/XmlIo.h>
 #include <MaterialXGenOgsXml/GlslFragmentGenerator.h>
 #include <MaterialXGenShader/Util.h>
 
-MaterialXData::MaterialXData(const std::string& materialXDocumentPath, 
-                             const std::string& elementPath, 
-                             const MaterialX::FileSearchPath& librarySearchPath)
-    : _shaderGenerator(mx::GlslFragmentGenerator::create()),
-      _genContext(_shaderGenerator),
-      _librarySearchPath(librarySearchPath)
+namespace
 {
-    setRenderableElement(materialXDocumentPath, elementPath);
+mx::DocumentPtr createDocument( const std::string& materialXDocumentPath,
+                                const MaterialX::FileSearchPath& librarySearchPath )
+{
+    // Create document
+    mx::DocumentPtr document = mx::createDocument();
+    if (!document)
+    {
+        throw mx::Exception("Failed to create a MaterialX document");
+    }
+
+    // Load libraries
+    static const mx::StringVec libraries = { "stdlib", "pbrlib", "bxdf", "stdlib/genglsl", "pbrlib/genglsl" };
+    MaterialXMaya::loadLibraries(libraries, librarySearchPath, document);
+
+    // Read document contents from disk
+    mx::XmlReadOptions readOptions;
+    readOptions.skipDuplicateElements = true;
+    mx::readFromXmlFile(document, materialXDocumentPath, mx::EMPTY_STRING, &readOptions);
+
+    return document;
+}
 }
 
-bool MaterialXData::setRenderableElement(const std::string& materialXDocument, const std::string& elementPath)
+MaterialXData::MaterialXData(   mx::DocumentPtr document,
```

Comment:
Providing a version of the constructor accepting a preexisting document, for situations when we want to edit something on the node without touching the document.